clock_gettime detection fails

After upgrading from 2.6.4 to 2.7.2 the build fails on my setup because HAVE_CLOCK_GETTIME is not properly detected.

The build fails with the following error
```
usr/include/time.h:336:12: error: conflicting types for 'clock_gettime'
 extern int clock_gettime (clockid_t __clock_id, struct timespec *__tp) __THROW;
```
because HAVE_CLOCK_GETTIME is not defined.


This is the reason why it fails
```
check_function_exists(clock_gettime HAVE_CLOCK_GETTIME)
if(HAVE_CLOCK_GETTIME)
	add_definitions(-DHAVE_CLOCK_GETTIME)
endif()

if(CMAKE_SYSTEM_NAME MATCHES "Linux")
	check_library_exists(rt clock_gettime "time.h" HAVE_CLOCK_GETTIME)
	if (HAVE_CLOCK_GETTIME)
		set(OPENSSL_LIBS ${OPENSSL_LIBS} rt)
	endif()
	set(OPENSSL_LIBS ${OPENSSL_LIBS} pthread)
endif()
```

After the first check (check_function_exists) the HAVE_CLOCK_GETTIME variable is set to false in the cmake cache. Then when clock_gettime is searched for in the rt library using check_library_exists, no check is performed since cmake already has a value in the cache for HAVE_CLOCK_GETTIME

Line 35 in
[https://github.com/Kitware/CMake/blob/master/Modules/CheckLibraryExists.cmake](https://github.com/Kitware/CMake/blob/master/Modules/CheckLibraryExists.cmake)

So the HAVE_CLOCK_GETTIME variable should be cleared befored the call to check_function_exists if it is false, otherwise no check is performed.
Something like
```
if (NOT HAVE_CLOCK_GETTIME)
    unset(HAVE_CLOCK_GETTIME CACHE)
endif ()
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -264,27 +264,32 @@ endif()
 
 set(OPENSSL_LIBS tls ssl crypto)
 
+# Add additional required libs
 if(WIN32)
 	set(OPENSSL_LIBS ${OPENSSL_LIBS} ws2_32)
 endif()
-
-check_function_exists(clock_gettime HAVE_CLOCK_GETTIME)
-if(HAVE_CLOCK_GETTIME)
-	add_definitions(-DHAVE_CLOCK_GETTIME)
-endif()
-
 if(CMAKE_SYSTEM_NAME MATCHES "Linux")
 	set(OPENSSL_LIBS ${OPENSSL_LIBS} pthread)
-	check_library_exists(rt clock_gettime "time.h" HAVE_CLOCK_GETTIME)
-	if (HAVE_CLOCK_GETTIME)
-		set(OPENSSL_LIBS ${OPENSSL_LIBS} rt)
-	endif()
 endif()
 if(CMAKE_SYSTEM_NAME MATCHES "HP-UX")
 	set(OPENSSL_LIBS ${OPENSSL_LIBS} pthread)
 endif()
 if(CMAKE_SYSTEM_NAME MATCHES "SunOS")
 	set(OPENSSL_LIBS ${OPENSSL_LIBS} nsl socket)
+endif()
+
+if(CMAKE_SYSTEM_NAME MATCHES "Linux")
+	# Check if we need -lrt to get clock_gettime on Linux
+	check_library_exists(rt clock_gettime "time.h" HAVE_CLOCK_GETTIME)
+	if (HAVE_CLOCK_GETTIME)
+		set(OPENSSL_LIBS ${OPENSSL_LIBS} rt)
+	endif()
+else()
+	# Otherwise, simply check if it exists
+	check_function_exists(clock_gettime HAVE_CLOCK_GETTIME)
+endif()
+if(HAVE_CLOCK_GETTIME)
+	add_definitions(-DHAVE_CLOCK_GETTIME)
 endif()
 
 check_type_size(time_t SIZEOF_TIME_T)

